p4include/psa.p4

```diff
@@ -29,69 +29,233 @@ limitations under the License.
  * These types need to be defined before including the architecture file
  * and the macro protecting them should be defined.
  */
-#define PSA_CORE_TYPES
-#ifdef PSA_CORE_TYPES
-/* The bit widths shown below are only examples.  Each PSA
- * implementation is free to use its own custom width in bits for
- * those types that are bit<W> for some W.  The only reason that there
- * are example numerical widths in this file is so that we can easily
- * compile this file, and example PSA P4 programs that include it. */
-
-typedef bit<10> PortId_t;
-typedef bit<10> MulticastGroup_t;
-typedef bit<10> CloneSessionId_t;
-typedef bit<3>  ClassOfService_t;
-typedef bit<14> PacketLength_t;
-typedef bit<16> EgressInstance_t;
-typedef bit<48> Timestamp_t;
+#define PSA_ON_BMV2_CORE_TYPES
+#ifdef PSA_ON_BMV2_CORE_TYPES
+/* The bit widths shown below are specific to the BMv2 psa_switch
+ * target.  These types do _not_ dictate what sizes these types should
+ * have for any other implementation of PSA.  Each PSA implementation
+ * is free to use its own custom width in bits for those types that
+ * are bit<W> for some W.  One reason they are here is to support the
+ * implementation of PSA on BMv2.  Another is so that we can easily
+ * compile this file, and example PSA P4 programs that include it.
+ *
+ * In many cases, the bit widths of these types have been chosen to be
+ * the same as corresponding types in the v1model architecture, but
+ * this is simply because it should help keep the differences between
+ * BMV2 simple_switch and BMv2 psa_switch a tiny bit smaller. */
+
+/* These are defined using `typedef`, not `type`, so they are truly
+ * just different names for the type bit<W> for the particular width W
+ * shown.  Unlike the `type` definitions below, values declared with
+ * the `typedef` type names can be freely mingled in expressions, just
+ * as any value declared with type bit<W> can.  Values declared with
+ * one of the `type` names below _cannot_ be so freely mingled, unless
+ * you first cast them to the corresponding `typedef` type.  While
+ * that may be inconvenient when you need to do arithmetic on such
+ * values, it is the price to pay for having all occurrences of values
+ * of the `type` types marked as such in the automatically generated
+ * control plane API.
+ *
+ * Note that the width of typedef <name>Uint_t will always be the same
+ * as the width of type <name>_t. */
+
+/* The comments after each line give the name of a similar field in
+ * the BMv2 simple_switch implementation, on which the width of this
+ * PSA type is based. */
+typedef bit<9>  PortIdUint_t;          // ingress_port, egress_port
```

Comment:
I wonder that for the sake of simplicity we should try to have these types match the P4Runtime ones for bmv2 PSA.
For example, when creating a bmv2 instance, one can choose arbitrary port numbers and there is no reason for the user not to choose the SDN port numbers (see P4Runtime spec). So we could just use `bit<32>` for ports and say that for bmv2 port translation is never required. This way the `PSA_PORT_CPU` can just match the P4Runtime one: 0xfffffffd (https://github.com/p4lang/p4runtime/blob/master/proto/p4/v1/p4runtime.proto#L696).

@jafingerhut don't you think it's something worth considering? Otherwise, if we go with a `PSA_PORT_CPU` which is different from the P4Runtime one for example, then the P4Runtime server implementation for bmv2 PSA is required to perform translation.

I think there is a lot of value in achieving consistency in this. We are less likely to confuse people :)

The only drawback potentially is that it may be harder to test port translation code on bmv2, since no translation is actually needed but I doubt it is an issue. One can always use an identity mapping...

BTW I took ports as an example, but this applies to the other types which appear in P4Runtime. I think the goal of consistency with P4Runtime should be more important than consistency with v1model.

Reply:
Sounds like a reasonable idea to me.  I will soon update the PR based on this suggestion...

Reply:
So the BMv2 psa_switch would need to handle 32-bit port numbers in the data path code?  That could make for some interesting psa_switch command lines, e.g.:
```
psa_switch --log-console -i 939280@veth2 -i 1004902@veth4
```

Reply:
bmv2 already supports arbitrary 32-bit port numbers, otherwise I would not have suggested it
I expect that "front-panel" ports will always be small integers. The CPU port & recirculate port are large integers but will be hardcoded in the code.
If someone uses some unusual port numbers, it is likely that they are simulating large topologies of bmv2 switches and orchestrating it through scripts.
It's much more likely to be `psa_switch --log-console -i 1@veth2 -i 2@veth4`

Reply:
PortId_t is 32 bits in P4Runtime, and ClassOfService_t is 8 bits I think.

Are there P4Runtime sizes specified for the other 5 types, that at least so far do not have a requirement for runtime numeric translation?

+ MulticastGroup
+ CloneSessionId
+ PacketLength
+ EgressInstance
+ Timestamp

Reply:
I would recommend to just use the width of the `InHeader` variant of the type for these. P4Runtime uses 32-bit number for all of them but for non-translated types, the value must fit in the data plane type and the `InHeader` variant is supposed to be an upper bound for all targets.